### DISCUSSION TITLE
Bug passing and identifying custom class in Godot 4b1 on macOS?

### DISCUSSION THREAD

**Discussion by Gatada** [godotengine/godot-proposals#5506] [Engine Core]
I am seeing unexpected behaviour when trying to filter a node based on class.

Let's say the idea is for the Iron Giant to only consume instances of `Vehicle`.

This is the `Vehicle` class I created:

```class_name Vehicle
extends RigidBody3D

@icon("res://assets/icon/VehicleCustom.svg")

func get_class():
    return "Vehicle"

func is_class(name_of_class):
    return name_of_class == "Vehicle"
```

Now, to identify the nodes that are instances of Vehicle, so I can "consume" them. After having identified the `body` as a `Vehicle` I even pass it as such - to ensure the recipient gets the required type:

So here the class "consumed" is RigidBody3D:

```func _entered_by(body) -> void:
    if body.is_class("Vehicle"):
        consume_vehicle(body as Vehicle)

func consume_vehicle(car: Vehicle) -> void: ## Entry condition requires that car is of type Vehicle
    print("Consuming ", car.get_class()) ## THIS HOWEVER PRINTS: Consuming RigidBody3D
```
 
Though like this - it returns Vehicle:

```func _entered_by(body) -> void:
    if body.is_class("Vehicle"):
        consume_vehicle(body as Vehicle)

func consume_vehicle(car) -> void: ## Now I accept any class instance
    print("Consuming ", car.get_class()) ## Resulting in: Consuming Vehicle
```

I did not expect this to happen? By passing a `Vehicle` the class is received as a `RigidBody3D`? I would maybe accept the other way around, but even that should be prevented - as the instance is passes as a `Vehicle`.

Is this a bug in Godot 4b1?

**Comment by Calinou** [comment]
> Is this a bug in Godot 4b1?

This is not the place to report (potential) bugs. Use <https://github.com/godotengine/godot>'s issue tracker instead.
